### PR TITLE
Update scroll_overlay

### PR DESCRIPTION
**Pull request by nt4f04uNd**
* Migrated to null-safety
* Migrated to embedding v2
* In Flutter ListView changed tracking speed from the only fling to any time it moves
* Android now also reports speed. For that I replaced the ListView with RecyclerView, because only latter exposes this API (or at least it's more precise and clear)
* Changed Android side font size to match the Flutter side - was 15, should be 16

<details>
<summary>Video before PR</summary>

https://user-images.githubusercontent.com/39104740/137771222-8dc48fdc-232f-4d68-86f1-d92e137d3bda.mp4
</details>

<details>
<summary>Video after PR</summary>

https://user-images.githubusercontent.com/39104740/137774736-2f2fefde-2e02-4578-995b-86798077d5a4.mp4
</details>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 1]
```diff
@@ -3,151 +3,190 @@
...
+                if (flutterEngine == null)
+                    return;
+                if (!locked) {
+                    velocitySink.success((double) currentDy * ticksPerSecond);
+                    if (currentDy == 0) {
+                        locked = true;
+                    }
+                    currentDy = 0;
+                }
+                handler.postDelayed(velocityTrackerRunnable,1000 / ticksPerSecond);
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:52]:
```suggestion
                handler.postDelayed(velocityTrackerRunnable, 1000 / ticksPerSecond);
```

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 2]
```diff
@@ -3,151 +3,190 @@
...
+        boolean locked = false;
 
-    private static final int[] OVERLAY_COLORS = new int[]{0x40ff0000, 0x4000ff00, 0x400000ff};
+        final int ticksPerSecond = 25;
+        final Handler handler = new Handler(Looper.getMainLooper());
+        Runnable velocityTrackerRunnable = new Runnable() {
+            @Override
+            public void run() {
+                if (flutterEngine == null)
+                    return;
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:44]:
```suggestion
                if (flutterEngine == null) {
                    return;
                }
```

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 3]
```diff
@@ -3,151 +3,190 @@
...
+    FlutterEngine flutterEngine;
+    EventChannel.EventSink velocitySink;
 
-    private static final String VELOCITY_CHANNEL = "scroll_overlay.flutter.io/velocity";
+    class VelocityTracker extends RecyclerView.OnScrollListener {
+        int currentDy = 0;
+        boolean locked = false;
 
-    private static final int[] OVERLAY_COLORS = new int[]{0x40ff0000, 0x4000ff00, 0x400000ff};
+        final int ticksPerSecond = 25;
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:38]: how is this value determined?

Comment (nt4f04uNd) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:38]: I picked it arbitrarily and it looked good. Do you have any concerns about this?

Comment (blasten) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:38]: maybe try explaining what this value does, and why 25 is a good value for it

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 4]
```diff
@@ -3,151 +3,191 @@
...
+                    }
+                }
+        );
 
-        ListView overlayList = findViewById(R.id.overlay_list);
-        overlayList.setDividerHeight(0);
-        overlayList.setAdapter(new OverlayAdapter());
+        // Create list
+        RecyclerView recyclerView = findViewById(R.id.recycler_view);
+        recyclerView.addItemDecoration(new DividerItemDecoration(this, 0)); // Remove divider
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:107]: what part is removing the divider?

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 5]
```diff
@@ -3,151 +3,191 @@
...
+        class OverlayViewHolder extends RecyclerView.ViewHolder {
+            public OverlayViewHolder(@NonNull View itemView) {
+                super(itemView);
+            }
+
+            void bind(int position) {
+                int color = OVERLAY_COLORS[position % OVERLAY_COLORS.length];
+                itemView.setBackground(new ColorDrawable(color));
+                final int height = (int) TypedValue
+                        .applyDimension(TypedValue.COMPLEX_UNIT_DIP, 40 + position, getResources().getDisplayMetrics());
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:179]: `40` was already in the code, but would you mind creating a constant for it that auto explains what it controls?

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java (hunk 4 above)
Comment (nt4f04uNd) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:107]: This line does, I know it's confusing https://stackoverflow.com/a/47119961/9710294

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java (hunk 3 above)
Comment (nt4f04uNd) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java]: done - renamed the variable and added comments both in native and dart sides

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java (hunk 5 above)
Comment (nt4f04uNd) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java]: done, both in dart and native

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java [hunk 6]
```diff
@@ -3,151 +3,205 @@
...
+                    return;
+                }
+                if (!locked) {
+                    velocitySink.success((double) currentDy * measurementsPerSecond);
+                    if (currentDy == 0) {
+                        locked = true;
+                    }
+                    currentDy = 0;
+                }
+                handler.postDelayed(velocityTrackerRunnable, 1000 / measurementsPerSecond);
```
Comment (blasten) [inline scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:60]: shouldn't this run on each frame?  e.g. Is it better to schedule this using https://developer.android.com/reference/android/view/Choreographer.FrameCallback ?

**Comment by blasten** [comment]
Also, cc @xster since you are original author of this app

File: scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java (hunk 6 above)
Comment (nt4f04uNd) [reply scroll_overlay/android/app/src/main/java/io/flutter/scroll_overlay/ScrollOverlayActivity.java:60]: We don't want it to update too often. The comment I left on `measurementsPerSecond` seems relevant here

```
* Setting this to not too small value - to get a meaningful velocity information,
* and not too big - to distinguish individual digits after thousands.
```

**Comment by Piinks** [comment]
It looks like there is some alignment drift between Android/Flutter list items, you can see the further away from the top you are, the dividing line drifts. Is this expected? It's hard to tell if we are lining up right when we fling far down the list if they are already potentially out of alignment.

**Comment by nt4f04uNd** [comment]
I'm not sure I understand the "drift", do you mean that item height increases? I think it was done intentionally. It is happening in both Flutter and Android sides - the divider is just an indicator for the Flutter side and you can see the same happening with Android colored tiles.

**Comment by Piinks** [comment]
Should the Android and Flutter items line up then if they are both the same size? At the top the divider lines up at the edge, but it doesn't the further down the screen you go.

**Comment by nt4f04uNd** [comment]
Ah, I see. If I drag slowly, I can see there's a so called touch slop on Android side ([kTouchSlop](https://api.flutter.dev/flutter/gestures/kTouchSlop-constant.html) in Flutter), but not in Flutter - this seems to be causing the drift. I'll take a look.

**Comment by nt4f04uNd** [comment]
Unfortunately, I don't think there's a way to configure that neither in Flutter, nor Android. Even if we add some additional gesture detectors to enable the slop in Flutter, the values are different between Flutter and Android.

The only options is using reflection - which I did in the last commit. @Piinks can you see if that's better?

**Comment by Piinks** [comment]
Thanks for investigating @nt4f04uNd! I think this looks good. This will be very helpful, thank you!

**Review by Piinks — APPROVED** [review]

**Comment by Hixie** [comment]
@nt4f04uNd is this ready to land?

**Pull request merged**